src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LocalProxyVarsGen.java

```diff
@@ -0,0 +1,322 @@
+ *  This code and its internal interfaces are subject to change or
+ *  deletion without notice.</b>
+ */
+public class LocalProxyVarsGen extends TreeTranslator {
+
+    protected static final Context.Key<LocalProxyVarsGen> valueInitializersKey = new Context.Key<>();
+
+    public static LocalProxyVarsGen instance(Context context) {
+        LocalProxyVarsGen instance = context.get(valueInitializersKey);
+        if (instance == null)
+            instance = new LocalProxyVarsGen(context);
+        return instance;
+    }
+
+    private final Types types;
+    private final Names names;
+    private final Target target;
+    private TreeMaker make;
+    private final UnsetFieldsInfo unsetFieldsInfo;
+
+    private ClassSymbol currentClass = null;
+    private JCClassDecl currentClassTree = null;
+    private MethodSymbol currentMethodSym = null;
+
+    private final boolean noLocalProxyVars;
+
+    @SuppressWarnings("this-escape")
+    protected LocalProxyVarsGen(Context context) {
+        context.put(valueInitializersKey, this);
+        make = TreeMaker.instance(context);
+        types = Types.instance(context);
+        names = Names.instance(context);
+        target = Target.instance(context);
+        unsetFieldsInfo = UnsetFieldsInfo.instance(context);
+        Options options = Options.instance(context);
+        noLocalProxyVars = options.isSet("noLocalProxyVars");
+    }
+
+    public JCTree translateTopLevelClass(JCTree cdef, TreeMaker make) {
+        if (!noLocalProxyVars) {
+            try {
+                this.make = make;
+                return translate(cdef);
+            } finally {
+                // note that recursive invocations of this method fail hard
+                this.make = null;
+            }
+        } else {
+            return cdef;
+        }
+    }
+
+    java.util.List<JCVariableDecl> strictInstanceFields;
+
+    @Override
+    public void visitClassDef(JCClassDecl tree) {
+        ClassSymbol prevCurrentClass = currentClass;
+        JCClassDecl prevCurrentClassTree = currentClassTree;
+        MethodSymbol prevMethodSym = currentMethodSym;
+        java.util.List<JCVariableDecl> prevStrictInstanceFields = strictInstanceFields;
+        try {
+            currentClass = tree.sym;
+            currentClassTree = tree;
+            currentMethodSym = null;
+            super.visitClassDef(tree);
+            strictInstanceFields = tree.defs.stream()
+                    .filter(t -> t.hasTag(VARDEF))
+                    .map(t -> (JCVariableDecl)t)
+                    .filter(vd -> vd.sym.isStrict() && !vd.sym.isStatic())
+                    .collect(List.collector());
+            if (!strictInstanceFields.isEmpty()) {
+                for (JCTree t : tree.defs) {
+                    if (t.hasTag(JCTree.Tag.METHODDEF)) {
+                        JCMethodDecl md = (JCMethodDecl) t;
+                        // ignore telescopic and generated constructors
```

Comment:
I think TreeInfo::hasAnyConstructorCall might check this for you?

Reply:
Uhmmm probably that doesn't deal with GENERATEDCONSTR (although -- a default constructor should not do much damage -- just maybe some redundant analysis?)

Reply:
well but TreeInfo::hasAnyConstructorCall will return true for `this()` which I want to ignore